cmds/wget/wget_test.go

```diff
@@ -103,42 +85,52 @@ func getFreePort(t *testing.T) int {
 
 // TestWget implements a table-driven test.
 func TestWget(t *testing.T) {
-	tmpDir, execPath := testutil.CompileInTempDir(t)
-	defer os.RemoveAll(tmpDir)
-
 	// Start a webserver on a free port.
-	port := getFreePort(t)
 	unusedPort := getFreePort(t)
-	h := handler{}
+
+	server, err := net.Listen("tcp", ":0")
+	if err != nil {
+		t.Fatalf("Cannot get free port for server: %v", err)
+	}
+	port := server.Addr().(*net.TCPAddr).Port
 	go func() {
-		t.Fatal(http.ListenAndServe(fmt.Sprintf(":%d", port), h))
+		t.Fatal(http.Serve(server, handler{}))
 	}()
 
-	time.Sleep(500 * time.Millisecond) // TODO: better synchronization
```

Comment:
This sleep is still needed, no? I remember before there is a race between the server starting and wget starting. wget does not retry if the port is closed. The sleep adds ample delay to ensure the server is started first.

Reply:
No? Once you listen, the port is open to be connected to.

Reply:
You are completely right. You moved `Listen` outside the goroutine, so this is perfect!